### ISSUE TITLE
Amesos2: Build error (complex, Amesos2_Umfpack_FunctionMap.hpp)

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1584]
@trilinos/amesos2 
Blocks: Sierra ticket 17124

@prwolfe writes: "`Amesos2_Umfpack_FunctionMap.hpp` includes `umfpack.hpp` and then promptly redeclares [the functions in that header file].  The main difference being `double *` for two arguments instead of `double[]`. Worse, the double complex uses functions not in our installation of umfpack.  This is the current bottleneck for getting a build."

It does look a little weird that the header file in question first includes `umfpack.hpp`, then does an extern declaration for functions in that header file.  Does it even need to include the header file if it has extern declarations?

**Comment by ndellingwood** [comment]
@mhoemmen thanks for diagnosing a potential fix; I'm about to start building to reproduce the error, to clarify if I build with Umfpack enabled and Complex enabled, I should be able to reproduce this?

**Comment by mhoemmen** [comment]
@ndellingwood This was with Sierra's "external Trilinos" build, which uses Trilinos' usual CMake configuration process with complex arithmetic enabled.  Thus, it should be possible for us to reproduce this.  Thanks for picking this up!

**Comment by ndellingwood** [comment]
@mhoemmen 
I configured Trilinos with Complex enabled and Umfpack enabled within Amesos2 and was able to build Trilinos (locally on my laptop). 
I enabled Complex with these configure options:
```
-D Teuchos_ENABLE_COMPLEX=ON \
-D Tpetra_INST_COMPLEX_DOUBLE=ON \
```

Any extra info you can pass my way to reproduce the issue they're seeing? Maybe a version of Umfpack they are using? 

It looks like this is the umfpack lib I have installed on my laptop: `libumfpack.a -> libumfpack.5.6.2.a`

Also, the `umfpack.hpp` header is needed to be declared in `Amesos2_Umfpack_FunctionMap.hpp` otherwise the build will break.

**Comment by srajama1** [comment]
@MicheldeMessieres : FYI and any comments would help.

**Comment by MicheldeMessieres** [comment]
I'm testing on Umfpack 5.7.6 at the moment. I think we may want to abolish all the extern function declarations and perhaps simply do this (either with or without an extern as that version at least has it built in):

namespace UMFPACK {
  #include "umfpack.h" 
}

or like Amesos (perhaps to support older includes that don't have extern in the include?)

namespace UMFPACK {
  extern "C" {
    #include "umfpack.h"
  }
}

All other extern declarations would be deleted in Amesos2_Umfpack_FunctionMap.hpp.

I just verified that will run/pass with complex - the use of the extern function declarations I think is a mistake left over from earlier type issues during the refactor. That will also fix the mismatched double * to double [] though I am not sure there is a difference in how the compiler handles those.

Note the Amesos2 Umfpack wrapper uses other things besides just the methods (UMFPACK_INFO, UMFPACK_CONTROL) so we will need the include somewhere.

The original amesos setup puts umfpack.h inside an extern so maybe older versions are setup differently and this should do the same. That I did not investigate yet.

It sounds like there was a mismatch for the format of the complex method based on the original comments. Can we verify which version was used and which versions we intend to support? That might be the issue blocking the build.

**Comment by mhoemmen** [comment]
@ndellingwood Sierra uses `Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON` instead of `Tpetra_INST_COMPLEX_DOUBLE=ON`.  Trilinos now prefers the former to the latter.  This should also give faster builds, since it makes kokkos-kernels prebuild the complex kernels (kokkos-kernels doesn't know about `Tpetra_INST_COMPLEX_DOUBLE`, but it _does_ know about `Trilinos_ENABLE_COMPLEX_DOUBLE`).

**Comment by mhoemmen** [comment]
@ndellingwood wrote:
> It looks like this is the umfpack lib I have installed on my laptop: libumfpack.a -> libumfpack.5.6.2.a

The version they build has header file markings that say "4.1".

**Comment by ndellingwood** [comment]
@mhoemmen thanks for the version number, do you know if it refers to SuiteSparse vs UMFPACK? 
I haven't been able to reproduce this yet, I tested against a local build of SuiteSparse 4.1.0 - here's quick update on what I tried:

* I downloaded and built SuiteSparse 4.1 and compiled Trilinos with Amesos2 with UMFPACK and tests+examples enabled, and Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON, but had no compiler errors. 
* I hacked some of the unit tests in Amesos2 to make sure some of the ComplexSolve tests were built and tested with Umfpack - all tests built and passed. 

If the 4.1 refers to UMFPACK itself, I'm having trouble tracking down a download of it - I looked in the ChangeLog in SuiteSparse and UMFPACK 4.1 was released in 2003, but the oldest version of SuiteSparse I can download is 2.1.1 which contains UMFPACK 5.0.1 (from 2006).

**Comment by mhoemmen** [comment]
@ndellingwood The header files in question have the year "2003" in them.  It looks like the stand-alone UMFPACK release.  Anyway, if getting rid of the `extern` declarations fixes it, that's already a good start :-D .

**Comment by ndellingwood** [comment]
Hi @mhoemmen, can provide a copy of the UMFPACK4.1 stand-alone code to myself and @MicheldeMessieres? 
The tests he ran removing the `extern` declarations was with a newer copy (5.7.6), and testing with 4.1 can verify if this addresses the issue that "double complex uses functions not in our installation of umfpack" 

I haven't been able to replicate any of the build errors, the oldest version of UMFPACK I've been able to test is 4.3, thanks to this link from @srajama1 http://ftp.mcs.anl.gov/pub/petsc/externalpackages/ 
With that standalone lib, all Amesos2 tests build and pass locally on my machine.

**Comment by srajama1** [comment]
This might be another option - Update the UMFPACK version in the user if that is acceptable to the users. Adding @rrdrake to see if has any comments as this version of UMFPACK being used is really old.

**Comment by rrdrake** [comment]
I couldn't exactly tell from the comments here .. would moving up the version of UMFPACK fix the compile error?

The only caveat to the UMFPACK version is that after SuiteSparse 3.0.0, UMFPACK changed to GPL, which Sierra cannot use.  The version of UMFPACK in SuiteSparse 3.0.0 is 5.1.0.

**Comment by MicheldeMessieres** [comment]
@ndellingwood Added PR #1591 which removes the extern declarations.

**Comment by srajama1** [comment]
@rrdrake : The upgrade will at least allow us to test the same versions. The older version of UMFPACK is not available anywhere I know. Another option is for you to give us the directory :)

I am trying to get permission for government use for Suitesparse even though it is GPL. Let us see how that goes.

**Comment by mhoemmen** [comment]
@ndellingwood Sierra has the version of UMFPACK in question; you could copy it out of sierra/code/TPLs_src.  Otherwise, I'm not sure how to get it.

**Comment by mhoemmen** [comment]
@ndellingwood Thanks for the phone conversation!  Please refer to Sierra ticket 17494 for discussion.

**Comment by ndellingwood** [comment]
Thanks @mhoemmen and @rrdrake for help getting a copy of UMFPACK 4.1. 
I tested locally with UMFPACK 4.1 with @MicheldeMessieres PR #1591 and separately with the develop branch of Trilinos, builds and tests all pass in both cases. 
Since the changes only affect Amesos2 with UMFPACK enabled, which is not tested in nightlies or by the checkin test script I'm assuming its safe for me to merge the PR without running the full checkin test.

**Comment by ndellingwood** [comment]
For reference issue #1255 contains discussion for adding UMFPACK support (and SuperLU, SuperLU-Dist) for nightly testing

**Comment by ndellingwood** [comment]
PR #1255 is merged. Tested locally with complex double on/off and umfpack 4.1.

**Comment by mhoemmen** [comment]
Thanks @ndellingwood ! :D

**Comment by ndellingwood** [comment]
@prwolfe I wanted to follow up and see if the merge PR was able to resolve the issue you were seeing?

**Comment by ndellingwood** [comment]
Unit tests pass with UMFPACK 4.1 and PR #1591; marking as closed, please reopen if necessary.

**Comment by prwolfe** [comment]
My current status as of yesterday.

Sierra continues to hunt for the umfpack symbols in a nonexistent UMFPACK namespace (declared in Amesos2_Umfpack_FunctionMap.hpp). To confirm I built Amesos2 in Trilinos with unpack enabled (4.1) and no modifications (the head of the develop branch). I got the exact same errors

../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_numeric(int const*, int const*, double const*, double const*, void*, void**, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_symbolic(int, int, int const*, int const*, double const*, void**, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_defaults(double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_defaults(double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_numeric(int const*, int const*, double const*, void*, void**, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_solve(int, int const*, int const*, double const*, double const*, double*, double*, double const*, double const*, void*, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_symbolic(int, int, int const*, int const*, double const*, double const*, void**, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_solve(int, int const*, int const*, double const*, double*, double const*, void*, double const*, double*)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_free_symbolic(void**)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_di_free_numeric(void**)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_free_numeric(void**)'
../../src/libamesos2.so.12.13: undefined reference to `UMFPACK::umfpack_zi_free_symbolic(void**)'
collect2: error: ld returned 1 exit status
make[2]: *** [packages/amesos2/test/solvers/Amesos2_KLU2_UnitTests.exe] Error 1
make[2]: Leaving directory `/scratch/prwolfe/trilinosBuild'
make[1]: *** [packages/amesos2/test/solvers/CMakeFiles/Amesos2_KLU2_UnitTests.dir/all] Error 2
make[1]: Leaving directory `/scratch/prwolfe/trilinosBuild'
make: *** [all] Error 2

I can't see how the linker can resolve these without rebuilding umfpack with significant source changes. I am working on a solution that simply moves the implementation into a cpp file without the namespace and leaves the current structs as declarations instead.

**Comment by mhoemmen** [comment]
It's not at all clear how the thing ever linked in the first place.  The header files get included inside of the `UMFPACK` namespace, but the UMFPACK implementation itself is C and does not use namespaces.  Could it be that y'all thought the UMFPACK TPL was enabled but it wasn't?  We know you tested but we're just mystified at how it could have worked.

**Comment by ndellingwood** [comment]
@mhoemmen 
I'll paste my Trilinos configuration below (some settings may have accumulated that are irrelevant for this issue). I built/tested with gcc 5.4. No build errors, unit tests passed, I also built/tested with example driver code and tested with umfpack as the solver (--solver=umfpack). I'm looking more closely at the unit tests now but still having issues reproducing. I'm using UMFPACK 4.1 provided by @rrdrake 

Any comments on configuration changes etc that may help reproduce would be helpful so I can help get this fixed as quickly as possible. 

```

#!/bin/bash

export SRCDIR=/Users/ndellin/Research/repos-duplicate/trilinos/Trilinos

MPI_HOME=/opt/local

BLAS_PATH=/Users/ndellin/Research/TPL/BLAS-3.6.0
LAPACK_PATH=/Users/ndellin/Research/TPL/lapack-3.6.1

SCOTCH_PATH=/Users/ndellin/Research/TPL/scotch_6.0.3
UMFPACK_PATH=/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic

export INSTALL_LOCATION=${PWD}/install

rm -rf CMakeCache.txt CMakeFiles

cmake \
\
-D CMAKE_INSTALL_PREFIX:PATH="${INSTALL_LOCATION}" \
-D CMAKE_CXX_FLAGS:STRING="-g" \
-D CMAKE_MAKE_PROGRAM="make" \
-D CMAKE_BUILD_TYPE:STRING=RELEASE \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
-D BUILD_SHARED_LIBS:BOOL=ON \
\
-D TPL_ENABLE_MPI:BOOL=ON \
  -D CMAKE_CXX_COMPILER:FILEPATH="`which mpicxx`" \
  -D CMAKE_C_COMPILER:FILEPATH="`which mpicc`" \
  -D CMAKE_Fortran_COMPILER:FILEPATH="`which mpifort`" \
\
-D TPL_ENABLE_ParMETIS:BOOL=OFF \
-D TPL_ENABLE_Scotch:BOOL=ON \
  -D Scotch_INCLUDE_DIRS:PATH="${SCOTCH_PATH}/include" \
  -D Scotch_LIBRARY_DIRS:PATH="${SCOTCH_PATH}/lib" \
  -D Scotch_LIBRARY_NAMES:STRING="scotch;scotcherr" \
-D TPL_ENABLE_BLAS:STRING=ON \
-D TPL_ENABLE_LAPACK:STRING=ON \
-D TPL_ENABLE_UMFPACK:STRING=ON \
  -D UMFPACK_INCLUDE_DIRS:PATH="${UMFPACK_PATH}/Include"\
  -D UMFPACK_LIBRARY_DIRS:PATH="${UMFPACK_PATH}/Lib"\
\
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=OFF \
-D Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_Amesos:BOOL=ON \
  -D Amesos_ENABLE_KLU:BOOL=ON \
  -D Amesos_ENABLE_UMFPACK:BOOL=ON \
  -D Amesos_ENABLE_TESTS:BOOL=ON \
  -D Amesos_ENABLE_EXAMPLES:BOOL=ON \
\
-D Trilinos_ENABLE_CHECKED_STL:BOOL=OFF \
-D Trilinos_ENABLE_OpenMP:BOOL=ON \
-D Trilinos_ENABLE_Kokkos:BOOL=ON \
  -D Kokkos_ENABLE_OpenMP:BOOL=ON \
  -D Kokkos_ENABLE_Serial:BOOL=ON \
  -D Kokkos_ENABLE_EXAMPLES:BOOL=OFF \
  -D Kokkos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-D Trilinos_ENABLE_Epetra=ON \
  -D Epetra_ENABLE_EXAMPLES:BOOL=OFF \
  -D Epetra_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EpetraExt=ON \
  -D EpetraExt_ENABLE_EXAMPLES:BOOL=OFF \
  -D EpetraExt_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_Tpetra:BOOL=ON \
  -D Tpetra_ENABLE_EXAMPLES:BOOL=OFF \
  -D Tpetra_ENABLE_TESTS:BOOL=OFF \
  -D Tpetra_INST_SERIAL:BOOL=ON \
-D Trilinos_ENABLE_TeuchosCore:BOOL=ON \
-D Trilinos_ENABLE_TeuchosRemainder:BOOL=ON \
-D Trilinos_ENABLE_ShyLUBasker=OFF \
-D Trilinos_ENABLE_Amesos2:BOOL=ON \
  -D Amesos2_ENABLE_EXAMPLES:BOOL=ON \
  -D Amesos2_ENABLE_TESTS:BOOL=ON \
  -D Amesos2_ENABLE_KLU2:BOOL=ON \
  -D Amesos2_ENABLE_TIMERS:BOOL=ON \
  -D Amesos2_ENABLE_Basker:BOOL=ON \
  -D Amesos2_ENABLE_UMFPACK:BOOL=ON \
${SRCDIR}
```

**Comment by MicheldeMessieres** [comment]
I think the links to umfpack lib work due to the extern "C" in umfpack.h which removes the name spacing for the linkage. So possibly the link errors are due to the umfpack lib not being present or working since umfpack is not part of the standard Trilinos SuiteSparse (setting of UMFPACK_INCLUDE_DIRS and UMFPACK_LIBRARY_DIRS)? Maybe having this include inside a namespace is not a great idea but I'm not sure if it breaks any standard. It was more of a cosmetic feature to preserve the patterns in Amesos2.

**Comment by MicheldeMessieres** [comment]
I tried a quick test and removed the namespace and all references to UMPACK:: in the file Amesos2_Umfpack_FunctionMap.hpp. It is not used elsewhere. If the namespace is causing an issue that might be a useful check. With or without the namespace is working for me.

**Comment by mhoemmen** [comment]
I don't mean to be rude, but are you sure that the UMFPACK TPL is actually enabled in your build?  I find it hard to believe that the code could possibly link if the header file is wrapped in a namespace.  Does your configure.out file show that the UMFPACK TPL is enabled?  Is `Amesos2_ENABLE_UMFPACK:BOOL=ON` in your CMakeLists.txt?

**Comment by mhoemmen** [comment]
I wonder also if there's something wrong with the CMake logic that makes `HAVE_AMESOS2_UMFPACK` not get defined.

**Comment by MicheldeMessieres** [comment]
Seems Amesos2_ENABLE_UMFPACK is ON and the TPL is enabled. I can confirm my Cmake says it is finding the Umpack lib. I'm running the Amesos2 tests and seeing the Umfpack test running. Earlier I had tried running my clang debug and stepping to the Umpack call lines just to be sure I was really getting there. I am just testing clang at this moment though.

**Comment by mhoemmen** [comment]
I'm on the phone with @ndellingwood right now trying to figure this out.  It seems like Trilinos was pointing to the wrong UMFPACK install.  @rrdrake is on his way to work with Nathan too.

**Comment by ndellingwood** [comment]
When looking at configure output the INCLUDE_DIRS paths were all set properly, but once I double checked my CMakeCache.txt I saw my tests earlier were using a libs from a system install of a newer Umfpack and not the those from the Library paths I specified (arrgh!). 

I rebuilt umfpack 4.1 (some directory restructured from config file above), did a clean start with configure using umfpack 4.1 this time and tried to build Trilinos, and encountered linkage errors.

In Umfpack 4.1, umfpack.h lacks the extern "C" that is in newer versions of Umfpack mentioned above - I verified this is present in SuiteSparse 2.1.1, the oldest version available online.

I wrapped the #include "umfpack.h" in Amesos2_Umfpack_FunctionMap.hpp with extern "C" and everything compiled and linked fine; however the Amesos2 unit test fails with Umfpack 4.1 and I'm looking into this now (Amesos Umfpack unit tests pass).

Here are some changes I had to make to my configure script to make sure the proper libraries were used (I had to enable AMD and set paths to that lib as well):
```
-D TPL_ENABLE_UMFPACK:STRING=ON \
  -D UMFPACK_INCLUDE_DIRS:PATH="${UMFPACK_PATH}/UMFPACK/Include;${UMFPACK_PATH}/AMD/Include"\
  -D UMFPACK_LIBRARY_DIRS:PATH="${UMFPACK_PATH}/UMFPACK/Lib;${UMFPACK_PATH}/AMD/Lib"\
  -D UMFPACK_LIBRARY_NAMES:STRING="umfpack;amd" \
-D TPL_ENABLE_AMD:STRING=ON \
  -D AMD_INCLUDE_DIRS:PATH="${UMFPACK_PATH}/AMD/Include"\
  -D AMD_LIBRARY_DIRS:PATH="${UMFPACK_PATH}/AMD/Lib"\
  -D AMD_LIBRARY_NAMES:STRING="amd" \
```


Here are relevant lines from my CMakeCache.txt line as verification/sanity check:

```
//No help, variable specified on the command line.
Amesos2_ENABLE_UMFPACK:BOOL=ON

//List of semi-colon separated paths to append to the compile invocations
// to find the headers for the TPL UMFPACK.  This is the final
// variable that is used in the build commands.  The user variable
// UMFPACK_INCLUDE_DIRS is used to look for the given headers first
// but is just a suggestion. This variable, however, is the final
// value and will not be touched.
TPL_UMFPACK_INCLUDE_DIRS:PATH=/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/UMFPACK/Include;/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/AMD/Include

//List of semi-colon separated full paths to the libraries for
// the TPL UMFPACK.  This is the final variable that is used in
// the link commands.  The user variable UMFPACK_LIBRARY_DIRS is
// used to look for the know library names but but is just a suggestion.
// This variable, however, is the final value and will not be touched.
TPL_UMFPACK_LIBRARIES:FILEPATH=/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/UMFPACK/Lib/libumfpack.a;/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/AMD/Lib/libamd.a

//List of semi-colon separated paths to append to the compile invocations
// to find the headers for the TPL AMD.  This is the final variable
// that is used in the build commands.  The user variable AMD_INCLUDE_DIRS
// is used to look for the given headers first but is just a suggestion.
// This variable, however, is the final value and will not be touched.
TPL_AMD_INCLUDE_DIRS:PATH=/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/AMD/Include

//List of semi-colon separated full paths to the libraries for
// the TPL AMD.  This is the final variable that is used in the
// link commands.  The user variable AMD_LIBRARY_DIRS is used to
// look for the know library names but but is just a suggestion.
// This variable, however, is the final value and will not be touched.
TPL_AMD_LIBRARIES:FILEPATH=/Users/ndellin/Research/TPL/UMFPACKv4.1/UMFPACK-fpic/AMD/Lib/libamd.a
```

**Comment by srajama1** [comment]
Does your tests work with newer UMFPACK and fail with 4.1 ?

**Comment by ndellingwood** [comment]
@MicheldeMessieres thanks for testing with the namespace removed and confirming there are no issues; would you mind submitting a PR with the changes you made, as well as wrapping #include "umfpack.h" with extern "C" ? This should resolve the issue for the older version of Umfpack currently needed by @prwolfe and @rrdrake

**Comment by ndellingwood** [comment]
Issue #1255 includes discussions about adding TPLs (including UMFPACK) to nightly testing. Mentioning here for tracking purposes.

**Comment by MicheldeMessieres** [comment]
@srajama1 I never tested the old version of umfpack myself. I'll get setup for that using your link.

**Comment by MicheldeMessieres** [comment]
@ndellingwood I'll remove the namespace and add the extern wrapping. I'll test this on the old and new umfpack versions and submit a new PR.

**Comment by ndellingwood** [comment]
@MicheldeMessieres thanks! I just emailed a copy of umfpack 4.1, I was never able to find this version online.

**Comment by MicheldeMessieres** [comment]
@@ndellingwood  Thanks! Was just about to ask about 4.1 versus 4.3.

**Comment by ndellingwood** [comment]
FYI the Amesos2 umfpack test that failed with Umfpack 4.1 passes with Umfpack 5.0.1 (from SuiteSparse 2.1.1)

**Comment by mhoemmen** [comment]
@prwolfe @rrdrake

**Comment by MicheldeMessieres** [comment]
In the newer versions of Umfpack there is support for passing the complex as a single array which is how it is currently used in the wrappers. So Ax is passed as the array (real and imaginary parts) and Az is set to NULL.

This is not supported in 4.1 or 4.3 which is why the complex tests currently fail (real is passing). An error is given that the parameter Az cannot be NULL. (Note I was testing 4.1 but did not test 4.3 yet, just basing this comment on the docs).

I would need to split the array up and pass real and imaginary components individually. I did not confirm this works yet but wanted to get feedback and make sure we still want to support those versions.

If we support the older format I'll also need to determine which version the change was actually implemented. In the documentation searching for the string "Future complex version" in the older versions shows a note that this was not implemented but being planned. I didn't investigate that further yet.

**Comment by rrdrake** [comment]
If it matters, the Sierra team would like to move to version 5.1 of UMFPACK in the near future.  We just need to schedule it in.  The current thrust was to update to a new Trilinos.

**Comment by srajama1** [comment]
@rrdrake : Do you care for complex support in 4.1 until you move to 5.1 ? Considering this is a new feature in Amesos2 I am guessing the answer is no. We can then disallow complex usage with 4.1 through Amesos2.

**Comment by rrdrake** [comment]
@srajama1 : As far as I know, we are not using the complex support in UMFPACK through Amesos2.

**Comment by srajama1** [comment]
We will disable it. If you need complex support we can look at supporting it for 4.1 or you can upgrade UMFPACK.

**Comment by ndellingwood** [comment]
@rrdrake @prwolfe @mhoemmen commit 8c43a0a9041635c462b4351081d57c12e2efa723 from PRs #1846 and #1849 is now in Trilinos develop branch (tested and pushed by checkin script); it includes the support needed as discussed in this issue and in PR comments (e.g. disabling complex testing for Umfpack version < 4.5).

**Comment by prwolfe** [comment]
I pushed an update to Sierra last week to pull us up to UMFPACK 5.1 - I think the only outstanding issue is getting UMFPACK testing onto the Trilinos dashboard.  I have that in the trilinos for sierra builds, but a tribits bug is blocking that right now (see https://github.com/TriBITSPub/TriBITS/issues/234).  Do we want to keep this issue open until it's done or close it and track it there?

**Comment by srajama1** [comment]
Looks like TriBITSPub/TriBITS#234 got closed as won't fix. TPL testing with Trilinos is better left to the framework team. @trilinos/framework . I am willing to close this, but I worry it will never get done as can be seen here from the lag.

**Comment by srajama1** [comment]
I am closing this. It doesn't help to keep it open for Amesos2 for a framework problem.
